File: 148.py
# Definition for singly-linked list.
# class ListNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution(object):
	def merge(self, h1, h2):
		dummy = tail = ListNode(0)
		while h1 and h2:
			if h1.val < h2.val:
				tail.next = h1
				tail = tail.next
				h1 = h1.next
			else:
				tail.next = h2
				tail = tail.next
				h2 = h2.next

		tail.next = h1 if h1 else h2
		return dummy.next


	def sortList(self, head):
		"""
		:type head: ListNode
		:rtype: ListNode
		"""
		if not head or not head.next:
			return head

		slow = fast = head
		pre = None
		while fast and fast.next:
			pre = slow
			fast = fast.next.next
			slow = slow.next

		pre.next = None
		return self.merge(*map(self.sortList, [head, slow]))

File: shadow/148.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) { val = x; }
 * }
 */
class Solution {
    public ListNode sortList(ListNode head) {
        if (head == null || head.next == null) {
            return head;
        }

        ListNode mid = findMid(head);
        ListNode right = sortList(mid.next);
        mid.next = null;
        ListNode left = sortList(head);

        return merge(left, right);
    }

    private ListNode findMid(ListNode head) {
        ListNode slow = head;
        ListNode fast = head.next;

        while (fast != null && fast.next != null) {
            slow = slow.next;
            fast = fast.next.next;
        }

        return slow;
    }

    private ListNode merge(ListNode left, ListNode right) {
        ListNode dummy = new ListNode(0);
        ListNode tail = dummy;

        while (left != null && right != null) {
            if (left.val < right.val) {
                tail.next = left;
                left = left.next;
            }
            else {
                tail.next = right;
                right = right.next;
            }

            tail = tail.next;
        }

        if (left == null) {
            tail.next = right;
        }

        if (right == null) {
            tail.next = left;
        }

        return dummy.next;
    }
}
